src/main/scala/contadamination/ContaminationModel.scala

```diff
@@ -0,0 +1,179 @@
+package contadamination
+
+import com.twitter.algebird.{BloomFilterMonoid, BloomFilter, BF}
+import contadamination.ContaminationModel.{ContaminationStats, FilterContext}
+import org.apache.spark.mllib.rdd.RDDFunctions
+import org.apache.spark.rdd.RDD
+import org.bdgenomics.adam.rdd.ADAMContext
+import org.bdgenomics.formats.avro.AlignmentRecord
+
+import scala.collection
+
+object ContaminationModel {
+
+  val defaultWinSize = 32
+  val defaultFpr = 0.1
+  val defaultSeed = 0
+
+  case class FilterContext(
+                name: String,     // Contig name
+                path: String,     // Reference path
+                numEntries: Int,  // Number of sliding windows
+                fpr: Double,      // False Positive Rate
+                bloomFilter: BF            // Bloom filter
+  )
+
+  case class ContaminationStats(
+                numReads: Int,    // Total number of reads checked
+                tn: Int,          // True Negatives (number of rejected reads)
+                rate: Double      // Contamination rate as the number of rejected reads over the total number of reads
+  )
+  
+  /**
+   * Build a ContaminationModel from a sequence of paths
+   * @param paths Sequence of FASTA files for the reference contigs
+   * @param winSize Sliding window size
+   * @param fpr False Positive Rate
+   * @param adamContext ADAM Context
+   * @return a ContaminationModel
+   */
+  def apply(paths: Seq[String], winSize: Int = defaultWinSize, fpr: Double = defaultFpr)(implicit adamContext: ADAMContext): ContaminationModel = {
+    paths.map(ContaminationModel(_, winSize, fpr)).reduce((x, y) => x ++ y)
+  }
+  
+  /**
+   * Build a ContaminationModel from a path
+   * @param path FASTA file with the reference contig
+   * @param winSize Sliding window size
+   * @param fpr False Positive Rate for the BloomFilter
+   * @param adamContext ADAM Context
+   * @return a ContaminationModel
+   */
+  def apply(path: String, winSize: Int = defaultWinSize, fpr: Double = defaultFpr)(implicit adamContext: ADAMContext): ContaminationModel = {
+    val fragments = adamContext.loadFasta(path, winSize).cache()
+    require(fragments.count() > 0, "At least one fragment is required")
+
+    // TODO check that there is only one contigName per file (right now we assume there is only one contig per file)
+    val contigName = fragments.first().getContig.getContigName
+
+    val slidingFragments = new RDDFunctions(fragments).sliding(2).flatMap({
```

Comment:
Lets assume our reads are of size 3, so we need to build the bloom filter using sequences of the same length. Let's say that the contig has a length of 16 nucleotides. We load the fasta file using ADAM which forces us to choose a fragment size, so I choose a size for the fragments of the same size as the read, 3 nucleotides per fragment, which means that I will end having an RDD of 6 fragments. Now I need to build the bloom filter with all the sliding windows of size 3 for the contig (which is splitted in fragments of size 3). So my algorithm does the following:

Contig = 0123456789ABCDEF
RDD = [012,345,678,9AB,CDE,F]

The RDDFunctions.sliding(2) gives me a sequence like [(012,345),(345,678),(678,9AB),(9AB,CDE),(CDE,F),(F)].
Then for each pair of fragments (012,345) I join both fragments "012345" and generate sliding sequences like [012, 123, 234]. But the last fragment of the RDD doesn't have to be multiple of the window size so I fill it with a non nucleotide character (for example the dot "."), in the example it would be something like (CDE,F) --> "CDEF.." --> [CDE, DEF, EF.], and the last (F) is just treated as "F.." for the bloom filter.

An alternative would be to just skip these incomplete fragments "EF." and "F..".

If you try to do this with other sliding sizes or fragment sizes the algorithm is more complex then.